module/zfs/qat_compress.c

```diff
@@ -0,0 +1,401 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+ * or http://www.opensolaris.org/os/licensing.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+#if defined(_KERNEL) && defined(HAVE_QAT)
+#include <linux/slab.h>
+#include <linux/vmalloc.h>
+#include <linux/pagemap.h>
+#include <linux/completion.h>
+#include <sys/zfs_context.h>
+#include "qat_compress.h"
+
+#define	TIMEOUT_MS		500 /* 0.5 seconds */
+#define	INST_NUM		6
+#define	GZIP_HEAD_SZ		2
+#define	GZIP_FOOT_SZ		4
+#define	QAT_MIN_BUF_SIZE	4096
+
+static CpaInstanceHandle dc_inst_handles[INST_NUM];
+static CpaDcSessionHandle session_handles[INST_NUM];
+static Cpa16U num_inst = 0;
+static Cpa16U inst = 0;
+int qat_init_done = 0;
+int zfs_qat_disable = B_FALSE;
+
+#define	PHYS_CONTIG_ALLOC(pp_mem_addr, size_bytes)	\
+	mem_alloc_contig((void *)(pp_mem_addr), (size_bytes))
+
+#define	PHYS_CONTIG_FREE(p_mem_addr)	\
+	mem_free_contig((void *)&(p_mem_addr))
+
+static inline struct page *mem_to_page(void *addr)
+{
+	if (!is_vmalloc_addr(addr))
+		return (virt_to_page(addr));
+
+	return (vmalloc_to_page(addr));
+}
+
+static void qat_dc_callback(void *p_callback, CpaStatus status)
+{
+	if (NULL != p_callback)
+		complete((struct completion *)p_callback);
+}
+
+static inline CpaStatus mem_alloc_contig(void **pp_mem_addr,
+						Cpa32U size_bytes)
+{
+	*pp_mem_addr = kmalloc_node(size_bytes, GFP_KERNEL, 0);
+	if (NULL == *pp_mem_addr)
+		return (CPA_STATUS_RESOURCE);
+	return (CPA_STATUS_SUCCESS);
+}
+
+static inline void mem_free_contig(void **pp_mem_addr)
+{
+	if (NULL != *pp_mem_addr) {
+		kfree(*pp_mem_addr);
+		*pp_mem_addr = NULL;
+	}
+}
+
+int
+qat_init(void)
+{
+	CpaStatus status = CPA_STATUS_SUCCESS;
+	Cpa32U sess_size = 0;
+	Cpa32U ctx_size = 0;
+	Cpa32U buff_meta_size = 0;
+	CpaDcSessionSetupData sd = {0};
+	int i;
+
+	if (zfs_qat_disable == B_TRUE || qat_init_done != 0)
+		return (0);
```

Comment:
Good point, I've removed the condition "zfs_qat_disable == B_TRUE" here, so in the load time the QAT will be loaded always (if the build option contains QAT), but at runtime check the QAT is disabled if zfs_qat_disable is set.